pm:device-constraint: Add power management constraints to devices

ceolin opened the pull request:
Allow to set a pm constraint on a device. When a device has a pm constraint it won't change its state until the constraint be released.
Device pm API will return `-EPERM` when the device has a constraint.

v2:
 - Fix commit messages
 -  Add test for this API
 
v3:
 - Change API namespace to `pm_device_state_lock*`
 - No longer reference counted (save memory using `device->pm->flags`)
 - "lock" a device state is only allowed when device pm runtime is disabled
 - Split the API test in device_runtime and pm_mgmt tests

Commit:
pm: device: Add device pm state lock

Add a new API to lock a device pm state.

When the device has its state locked, the kernel will no longer
suspend / resume devices when the system goes to sleep and device
runtime power management operations will fail.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Account device pm state lock

Do not execute pm operations on devices that the device pm state is
locked.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull reviewed (commented):
I'm not sure if I get the purpose of this change (maybe `constraing` naming doesn't help). Can potential usecases be described?

Commit:
tests: pm: Test pm device state lock

Test pm device state lock API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
> I'm not sure if I get the purpose of this change (maybe `constraing` naming doesn't help). Can potential usecases be described?

There are situations where a device is shared between a host cpu and a firmware in the platform.Then the firmware (Zephyr) is responsible for handling interruptions and forwarding them to Host, if unknown to the firmware of if a higher level software needs to handle them. In this scenario the host may set a constraint in this device to ensure that the zephyr pm subsystem does no interfer in this use case.

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -28,13 +28,18 @@ LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
 	int ret = 0;
 	struct pm_device *pm = dev->pm;
 
+	if (pm_device_constraint_check(dev)) {
+		return -EPERM;
+	}
```
I'm a bit skeptical about this change, since it makes drivers error handling more complicated. If a driver now does:

```
ret = pm_device_runtime_put(dev);
if (ret < 0) {
    return ret;
}
```

it will return an error. However, the driver operation will likely be ok (just that PM suspend/resume is not allowed at this time). To summarize, unless driver ignores `-EPERM`, it will become unusable. A similar situation happens in other scenarios, for which I've created https://github.com/zephyrproject-rtos/zephyr/pull/40670.

gmarull replied:
Another point: for runtime operation the device usage count may still be updated (but no action taken). I'm worried about introducing a mechanism that can potentially interfere with the runtime API usage, potentially leaving the system into an inconsistent state.

ceolin replied:
I thought about it, my opinion *now* is that to allow  increase/decrease the reference count when it does not trigger a state change makes this API behavior trickier.
My initial idea was to allow use it only when device runtime is disabled, the problem is that when runtime pm is disabled the device is always resumed, consequently if one wants to set a constraint (or lock) the device when the state is SUSPENDED it will be necessary to use `pm_device_action_run` and we don't want this API directly called. Though it makes the behavior much more predictable and the only `pm_device_runtime_enable` should be changed to return `-EPERM` when the device has a constraint (or is locked)

gmarull commented on tests/subsys/pm/power_mgmt/src/main.c:
```diff
@@ -366,6 +379,30 @@ void test_device_order(void)
+	pm_device_constraint_set((struct device *)device_a);
+	zassert_true(pm_device_constraint_check(device_a), NULL);
+
+	ret = pm_device_runtime_put(device_a);
```
runtime API behavior should be tested in the runtime API test

ceolin replied:
yep, had put here because could check it in both scnearios, system and runtime pm. Since you are removing runtime pm from this test I'll move it as well.

gmarull commented on include/pm/device.h:
```diff
@@ -424,8 +431,59 @@ static inline bool pm_device_wakeup_is_capable(const struct device *dev)
+ *
+ * @param dev Device instance.
+ */
+void pm_device_constraint_set(struct device *dev);
```
note that "constraint" is a really generic word, here it means that device state is _locked_, so I'd choose a better API naming.

ceolin replied:
agree, I just don't know if locked is much better, I was thinking that enable / disable was more appropriate but then it will be confuse with other APIs.

ceolin replied:
hmm, `pm_device_state_lock()`, `pm_device_state_unlock()` and `pm_device_state_is_locked()` sounds good though.

gmarull commented on include/pm/device.h:
```diff
@@ -424,8 +431,59 @@ static inline bool pm_device_wakeup_is_capable(const struct device *dev)
+ * @retval true If device has a constraint.
+ * @retval false If device does not have a constraint.
+ */
+bool pm_device_constraint_check(const struct device *dev);
```
`_is_set`?

ceolin replied:
changed API namespace to `state_lock*`, so this function now is called `pm_device_state_is_locked`

Commit:
tests: pm: Test state lock with runtime

Add device pm state lock to device runtime test.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

teburd commented on include/pm/device.h:
```diff
@@ -407,6 +410,44 @@ bool pm_device_wakeup_is_enabled(const struct device *dev);
+ * This function locks the current device state. After the state be
+ * locked the system power management and device runtime power
+ * management will not longer change the device state until the state
+ * be unlocked.
```
A little rewording makes this clearer possibly
"This function locks the current device power state. Once locked the device state
will not be change by system power management or device runtime power management until unlocked."

ceolin replied:
Indeed, much better with the change.

teburd commented on tests/subsys/pm/device_runtime_api/src/main.c:
```diff
@@ -190,6 +191,35 @@ static void test_api(void)
+	zassert_equal(locked, true, NULL);
+
+	/* Enable device runtime should fail because of the state lock */
+	pm_device_runtime_enable(dev);
```
I believe this is a fallible function in @gmarull's pr #40670 which this could benefit from perhaps

gmarull commented on include/pm/device.h:
```diff
@@ -298,6 +298,7 @@ const char *pm_device_state_str(enum pm_device_state state);
  * @retval -EALREADY If device is already at the requested state.
  * @retval -EBUSY If device is changing its state.
  * @retval -ENOSYS If device does not support PM.
+ * @retval -EPERM If device has a constraint.
```
`device has state locked`

gmarull commented on tests/subsys/pm/device_runtime_api/src/main.c:
```diff
@@ -190,6 +191,30 @@ static void test_api(void)
+	ret = pm_device_runtime_put(dev);
+	zassert_equal(ret, -EPERM, NULL);
+
+	locked = pm_device_state_is_locked(dev);
+	zassert_equal(locked, true, NULL);
```
this check can be removed, it's not relevant for this test (this test is just checking runtime behaves correctly when state is locked, not the lock API)

ceolin replied:
I don't agree here, if the runtime operation failed and the state is not locked, we have a problem and the test will catch it. Otherwise the put may fail (not because the state is locked) and we will accept.

gmarull replied:
My point is: `device_runtime_api` tests the runtime API behavior. If the device_runtime_api behavior depends on `pm_device_state_is_locked`, then the runtime tests needs to be updated. However, if `pm_device_state_is_locked` behavior depends on the runtime API, then that should be part of a separate test that checks `pm_device_state*` behavior. According to final changeset, the runtime PM API is not affected by this new API, so runtime tests don't need to be updated. I know some PM tests mix responsibilities, but that should be gradually cleaned up, otherwise it's really difficult to understand the purpose of tests.

ceolin replied:
yep, that is right. Was thinking in the initial behavior.

gmarull commented on tests/subsys/pm/device_runtime_api/src/main.c:
```diff
@@ -190,6 +191,30 @@ static void test_api(void)
+
+	pm_device_state_unlock(dev);
+
+	locked = pm_device_state_is_locked(dev);
+	zassert_equal(locked, false, NULL);
```
ditto

gmarull commented on include/pm/device_runtime.h:
```diff
@@ -42,7 +42,6 @@ void pm_device_runtime_enable(const struct device *dev);
  *
  * @retval 0 If the device runtime PM is disabled successfully.
  * @retval -ENOSYS If the functionality is not available.
- * @retval -EPERM If the device pm state is locked.
```
since change is added, then removed, can initial commit be fixed up so that the only the final behavior is introduced in this PR? It's now a bit confusing.

ceolin replied:
ok, will sqaush them,.

gmarull commented on include/pm/device.h:
```diff
@@ -418,6 +419,8 @@ bool pm_device_wakeup_is_capable(const struct device *dev);
  * system power management or device runtime power
  * management until unlocked.
  *
+ * @note The given device should not have device runtime enabled.
+ *
```
note: since there is this contract, this API call can fail.

Commit:
pm: device_runtime: Return possible error on enable

Change the function pm_device_runtime_enable() to return 0 on
success or an error code in case of error.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on doc/guides/pm/device_runtime.rst:
```diff
@@ -154,7 +154,10 @@ must perform the necessary operations to suspend the device.
         /* enable device runtime power management */
-        pm_device_runtime_enable(dev);
+        ret = pm_device_runtime_enable(dev);
+        if (ret != 0) {
```
in general `ret < 0` is used in the docs

ceolin replied:
positive value is wrong though.

gmarull replied:
well, API call claims to return a discrete set of negative values as error codes, so checking `< 0` covers all error scenarios (what this branch is testing for).

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -201,6 +203,7 @@ void pm_device_runtime_enable(const struct device *dev)
 
 end:
 	SYS_PORT_TRACING_FUNC_EXIT(pm, device_runtime_enable, dev);
```
nit: SYS_PORT_TRACING_FUNC_EXIT should also be updated to collect return code (and other related definitions such as `sys_port_trace_pm_device_runtime_enable_exit`).

gmarull commented on subsys/tracing/sysview/tracing_sysview.h:
```diff
@@ -639,8 +639,9 @@ void sys_trace_k_thread_info(struct k_thread *thread);
 #define sys_port_trace_pm_device_runtime_enable_enter(dev)		       \
 	SEGGER_SYSVIEW_RecordU32(TID_PM_DEVICE_RUNTIME_ENABLE,		       \
 				 (uint32_t)(uintptr_t)dev)
-#define sys_port_trace_pm_device_runtime_enable_exit(dev) \
-	SEGGER_SYSVIEW_RecordEndCall(TID_PM_DEVICE_RUNTIME_ENABLE)
+#define sys_port_trace_pm_device_runtime_enable_exit(dev, ret)		\
+	SEGGER_SYSVIEW_RecordEndCall(TID_PM_DEVICE_RUNTIME_ENABLE,	\
+				     (uint32_t)ret)
 #define sys_port_trace_pm_device_runtime_disable_enter(dev)		       \
 	SEGGER_SYSVIEW_RecordU32(TID_PM_DEVICE_RUNTIME_DISABLE,		       \
 				 (uint32_t)(uintptr_t)dev)
```
nit: for consistency with others, push all `\` to 80cols

ceolin replied:
it is mixing tabs with spaces ...

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -195,7 +201,9 @@ void pm_device_runtime_enable(const struct device *dev)
 	}
 
+end:
 	SYS_PORT_TRACING_FUNC_EXIT(pm, device_runtime_enable, dev);
```
```suggestion
	SYS_PORT_TRACING_FUNC_EXIT(pm, device_runtime_enable, dev, ret);
```

gmarull commented on tests/subsys/pm/power_mgmt/src/main.c:
```diff
@@ -304,7 +319,8 @@ void test_power_state_trans(void)
 
-	pm_device_runtime_enable(device_dummy);
+	ret = pm_device_runtime_enable(device_dummy);
+	zassert_true(ret == 0, "Failed to enable device runtime PM");
```
```suggestion
	zassert_equal(ret, 0, "Failed to enable device runtime PM");
```

ceolin replied:
it is consistent with the previous check
